### ISSUE TITLE
Need help writing 'serve' exec file for custom pytorch container

### ISSUE THREAD

**Issue by Harathi123** [aws/sagemaker-pytorch-container#45]
Hi,

I created custom container for sagemaker pytorch and was able to create training job. But while deploying the model, i am getting 'ping health check' error and didnt find anything in logs.

I included model_fn(), input_fn(), predict_fn(), output_fn() in training script. But I dont understand how to write 'serve' exec file. I think this is where i am going wrong. I refered to https://github.com/aws/sagemaker-pytorch-container/blob/master/src/sagemaker_pytorch_container/serving.py to write serve file and is as follows:
I am getting "no module named 'sagemaker_containers' error". 
It would be great if you can give some suggestions how to write serve file....

    #!/usr/bin/env python3

    from __future__ import absolute_import

    import logging
    from sagemaker_containers.beta.framework import (content_types, encoders, env, modules, transformer, worker)

    import torch

    logger = logging.getLogger(__name__)

    def default_model_fn(model_dir):
        """Loads a model. For PyTorch, a default function to load a model cannot be provided.
        Users should provide customized model_fn() in script.
        Args:
            model_dir: a directory where model is saved.
        Returns: A PyTorch model.
        """
        return transformer.default_model_fn(model_dir)

    def default_input_fn(input_data, content_type):
        """A default input_fn that can handle JSON, CSV and NPZ formats.
        Args:
            input_data: the request payload serialized in the content_type format
            content_type: the request content_type
        Returns: input_data deserialized into torch.FloatTensor or torch.cuda.FloatTensor depending if cuda is available.
        """
        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        np_array = encoders.decode(input_data, content_type)
        tensor = torch.FloatTensor(
            np_array) if content_type in content_types.UTF8_TYPES else torch.from_numpy(np_array)
        return tensor.to(device)

    def default_predict_fn(data, model):
        """A default predict_fn for PyTorch. Calls a model on data deserialized in input_fn.
        Runs prediction on GPU if cuda is available.
        Args:
            data: input data (torch.Tensor) for prediction deserialized by input_fn
            model: PyTvorch model loaded in memory by model_fn
        Returns: a prediction
        """
        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        model.to(device)
        input_data = data.to(device)
        model.eval()
        with torch.no_grad():
            output = model(input_data)
        return output

    def default_output_fn(prediction, accept):
        """A default output_fn for PyTorch. Serializes predictions from predict_fn to JSON, CSV or NPZ 
    format.
        Args:
            prediction: a prediction result from predict_fn
            accept: type which the output data needs to be serialized
        Returns: output data serialized
        """
        if type(prediction) == torch.Tensor:
            prediction = prediction.detach().cpu().numpy()
    
        return worker.Response(encoders.encode(prediction, accept), accept)

    def _user_module_transformer(user_module):
        model_fn = getattr(user_module, 'model_fn', default_model_fn)
        input_fn = getattr(user_module, 'input_fn', default_input_fn)
        predict_fn = getattr(user_module, 'predict_fn', default_predict_fn)
        output_fn = getattr(user_module, 'output_fn', default_output_fn)

        return transformer.Transformer(model_fn=model_fn, input_fn=input_fn, predict_fn=predict_fn,
                                   output_fn=output_fn)

    app = None

    def main(environ, start_response):
        global app
        if app is None:
            serving_env = env.ServingEnv()
            user_module = modules.import_module(serving_env.module_dir, serving_env.module_name)
            user_module_transformer = _user_module_transformer(user_module)
            user_module_transformer.initialize()
            app = worker.Worker(transform_fn=user_module_transformer.transform,
                            module_name=serving_env.module_name)

        return app(environ, start_response)

Thanks,
Harathi

**Comment by ChoiByungWook** [comment]
Hello,

How did you create your custom PyTorch container? Did you base it off of our Dockerfile or did you built it entirely from scratch?

The ping invocation error is due to https://docs.aws.amazon.com/sagemaker/latest/dg/your-algorithms-inference-code.html#your-algorithms-inference-algo-ping-requests. Our containers provides a hosting solution to satisfy the ping requests.

The "no module named 'sagemaker_containers' error" is because most likely your container doesn't have https://pypi.org/project/sagemaker-containers/ installed, which the SageMaker PyTorch container code utilizes.

**Comment by Harathi123** [comment]
Hi @ChoiByungWook , thanks for replying.

By the way the issue got resolved. Initially, i created custom container from scratch (installing python, pytorch etc.,). Now I found 'extending pytorch containers' in aws sagemaker examples. I followed it and created image from pytorch container which resolved the issue.

Thanks again,
Harathi

**Comment by ChoiByungWook** [comment]
Awesome!

Glad to hear! I'll close this ticket as it seems like your problem is resolved.

For other users that stumble upon this. The 'extending pytorch containers' is in reference to...
- https://github.com/awslabs/amazon-sagemaker-examples/blob/master/advanced_functionality/pytorch_extending_our_containers/pytorch_extending_our_containers.ipynb
- https://github.com/awslabs/amazon-sagemaker-examples
